mlflow/pipelines/utils/execution.py

```diff
@@ -270,7 +274,68 @@ def _get_step_output_directory_path(execution_directory_path: str, step_name: st
     )
 
 
-def _run_make(execution_directory_path, rule_name: str, extra_env: Dict[str, str]) -> None:
+class _ExecutionPlan:
+
+    _MSG_REGEX = r"^\t+# Running step: (\w+)\n$"
+    _FORMAT_STEPS_CACHED = "%s: No changes. Skipping."
+
+    def __init__(self, rule_name, output_lines_of_make: List[str], pipeline_steps: List[BaseStep]):
+        pipeline_step_names = [step.name for step in pipeline_steps]
+
+        steps_to_run = self._parse_output_lines(output_lines_of_make)
+        self.steps_cached = self._infer_cached_steps(rule_name, steps_to_run, pipeline_step_names)
+
+    @staticmethod
+    def _parse_output_lines(output_lines_of_make: List[str]) -> List[str]:
+        """
+        Parse the output lines of Make to get steps to run.
+        """
+
+        def get_step_to_run(output_line: str):
+            m = re.search(_ExecutionPlan._MSG_REGEX, output_line)
+            return m.group(1) if m else None
+
+        return list(
+            filter(
+                lambda step: step is not None,
+                (get_step_to_run(output_line) for output_line in output_lines_of_make),
+            )
```

Comment:
Nit: We can use a generator here instead of filter + lambda (I think lint will complain)